src/source_file.rs

```diff
@@ -0,0 +1,17 @@
+use miette::NamedSource;
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct SourceFile<'s> {
+    pub contents: &'s str,
+    pub filename: &'s str,
+}
+
+impl<'s> SourceFile<'s> {
+    pub fn new(contents: &'s str, filename: &'s str) -> Self {
```

Comment:
Didn't you flip these arguments at some point? That makes more sense to me. Maybe even don't make a `new` so we can't mix them up tbh (like what happened in the main function).